### PR TITLE
Stricter checking of index for vec_as_index()

### PR DESCRIPTION
**Pull request by krlmlr**
For tibble.

Closes https://github.com/tidyverse/tibble/issues/649.

### CODE REVIEW

File: tests/testthat/test-slice.R [hunk 1]
```diff
@@ -1,9 +1,12 @@
-context("test-slice")
```
Comment (lionel-) [inline tests/testthat/test-slice.R]: Why this change?

File: src/slice.c [hunk 2]
```diff
@@ -559,6 +559,14 @@ static SEXP chr_as_index(SEXP i, SEXP names) {
 }
 
 SEXP vec_as_index(SEXP i, R_len_t n, SEXP names) {
+  if (OBJECT(i)) {
+    Rf_errorcall(R_NilValue, "`i` must be an integer, character, or logical vector, not an object.");
```
Comment (lionel-) [inline src/slice.c]: Probably better to use `vec_cast()` to integer? Or `vec_coerce_cast()`, so that only coercible types are allowed (e.g. not `c("1", "2")`).

File: src/slice.c [hunk 3]
```diff
@@ -559,6 +559,14 @@ static SEXP chr_as_index(SEXP i, SEXP names) {
...
+  }
+
+  if (vec_dim_n(i) != 1) {
```
Comment (lionel-) [inline src/slice.c]: 👍

File: tests/testthat/test-slice.R (hunk 1 above)
Comment (krlmlr) [reply tests/testthat/test-slice.R]: `context()` is no longer needed in recent testthat.

Comment (lionel-) [reply tests/testthat/test-slice.R]: Looks like context are now generated from the file names, cool.

**Comment by lionel-** [comment]
Thanks!
